src/xenia/kernel/kernel_state.cc

```diff
@@ -229,37 +229,53 @@ object_ref<XModule> KernelState::GetModule(const char* name, bool user_only) {
       return retain_object(user_module.get());
     }
   }
-  return nullptr;
+
+  // Module not found, try loading it
+  return LoadUserModule(name);
 }
 
 object_ref<XThread> KernelState::LaunchModule(object_ref<UserModule> module) {
   if (!module->is_executable()) {
     return nullptr;
   }
 
-  SetExecutableModule(module);
-  XELOGI("KernelState: Launching module...");
+  XELOGI("KernelState: Launching module %s...", module->path().c_str());
 
   // Create a thread to run in.
   // We start suspended so we can run the debugger prep.
-  auto thread = object_ref<XThread>(
-      new XThread(kernel_state(), module->stack_size(), 0,
-                  module->entry_point(), 0, X_CREATE_SUSPENDED, true, true));
-
-  // We know this is the 'main thread'.
+  xe::kernel::object_ref<XThread> thread = nullptr;
+
+  if (!module->is_dll_module()) {
+    // Not a DLL module, run entrypoint as normal
+    thread = object_ref<XThread>(new XThread(
+        kernel_state(), module->stack_size(), XThread::StartupType::Normal, 0,
+        module->entry_point(), 0, X_CREATE_SUSPENDED, true, true));
+  } else {
+    // Run entrypoint as DllMain, using module handle as start context
+    thread = object_ref<XThread>(
```

Comment:
So, usually DllMain is called from the same thread that called the loader, i.e. `<game code> -> LoadLibrary -> <DLL code>`.
I believe it's this way on the xbox as well - so changing this behavior could introduce race conditions in guest code.